llvm/labs/lab2/durandin_vladimir/vladimir_inline_pass.cpp

```diff
@@ -0,0 +1,165 @@
+      for (BasicBlock &CalleeBB : *Callee) {
+        BasicBlock *NewBB = BasicBlock::Create(
+            F.getContext(),
+            CalleeBB.getName() + ".inlined." + std::to_string(counter_inlined),
+            &F);
+        if (Callee->getEntryBlock().getName() == CalleeBB.getName())
+          CalleeEntryBasicBlockPointer = NewBB;
+        BlockMap[&CalleeBB] = NewBB;
+      }
+      InsertBB->getTerminator()->setSuccessor(0, CalleeEntryBasicBlockPointer);
+
+      counter_inlined++;
+
+      // Copy instructions from corresponding blocks
+      ValueToValueMapTy VMap;
+      for (BasicBlock &CalleeBB : *Callee) {
+        BasicBlock *NewBB = BlockMap[&CalleeBB];
+        for (Instruction &Inst : CalleeBB) {
+          IRBuilder<> Builder(
+              NewBB); // Create IRBuilder with specified basic block
+          Instruction *NewInst = Inst.clone(); // Clone the instruction
+          Builder.Insert(NewInst); // Add the cloned instruction to the end of
+                                   // the basic block
+          VMap[&Inst] = NewInst;
+        }
+      }
+
+      // Update branch instructions to target correct blocks
+      // for (Instruction &Branch : *SplitBB) {
+      for (BasicBlock &CalleeBB : F) {
+        for (Instruction &Branch : CalleeBB) {
+          if (auto *BI = dyn_cast<BranchInst>(&Branch)) {
+            for (size_t i = 0, e = BI->getNumSuccessors(); i != e; ++i) {
+              BasicBlock *Successor = BI->getSuccessor(i);
+              if (BlockMap[Successor] != nullptr) {
+                BI->setSuccessor(i, BlockMap[Successor]);
+              }
+            }
+          }
+        }
+      }
+
+      for (BasicBlock &CalleeBB : *Callee) {
+        BasicBlock *NewBB = BlockMap[&CalleeBB];
+        for (Instruction &Inst : *NewBB) {
+          RemapInstruction(&Inst, VMap, RF_IgnoreMissingLocals, nullptr,
+                           nullptr);
+        }
+      }
+
+      // Create IRBuilder and set its context
+      IRBuilder<> Builder(F.getContext());
+
+      // Traverse the basic blocks in Caller
+      for (BasicBlock &BB : F) {
+        // Create a copy of the instruction list for the current basic block
+        std::vector<Instruction *> Instructions;
+        for (Instruction &I : BB) {
+          Instructions.push_back(&I);
+        }
+
+        for (Instruction *I : Instructions) {
+          if (auto *RI = dyn_cast<ReturnInst>(I)) {
+            if (RI->getParent() != SplitBB) {
+              RI->eraseFromParent();
+
+              // Create an unconditional branch instruction using IRBuilder
+              Builder.SetInsertPoint(
+                  &BB); // Set the insertion point to the current basic block
+              Builder.CreateBr(
+                  SplitBB); // Create an unconditional branch to SplitBB
+            }
+          }
+        }
+      }
+
+      // Remove the call instruction
+      CI->eraseFromParent();
+    }
+    return PreservedAnalyses::all();
```

Comment:
Do you think all analyses (like CFG, loop analysis, etc.) info should be kept after such a change?